cms/djangoapps/contentstore/views/course.py

```diff
@@ -374,15 +376,36 @@ def course_filter(course_summary):
         # TODO remove this condition when templates purged from db
         if course_summary.location.course == 'templates':
             return False
-
         return has_studio_read_access(request.user, course_summary.id)
+
+    def active_course_filter(course_summary):
+        """
+        Return accessible, usable courses, filtering out archived courses if requested.
+        """
+        if split_archived and course_summary.has_ended():
+            return False
+        return course_filter(course_summary)
+
+    def archived_course_filter(course_summary):
+        """
+        Return accessible, usable, archived courses
+        """
+        if split_archived and course_summary.has_ended():
+            return course_filter(course_summary)
+        return False
+
     if org is not None:
         courses_summary = [] if org == '' else CourseOverview.get_all_courses(orgs=[org])
     else:
         courses_summary = modulestore().get_course_summaries()
-    courses_summary = six.moves.filter(course_filter, courses_summary)
+
+    active_courses = six.moves.filter(active_course_filter, courses_summary)
+    archived_courses = ()
+    if split_archived:
+        archived_courses = six.moves.filter(archived_course_filter, courses_summary)
```

Comment:
I found the logic here a bit hard to follow.  How about doing the filtering in two separate steps?  First, use the filter from the old code.  Then, only if `split_archived` is set, partition the list into active and archived courses.  This would disentangle the three different conditions (the original `course_filter`, whether `split_archived` is set, whether a course is archived), so I feel it would become easier to understand.
```
active_courses = six.moves.filter(course_filter, courses_summary)
archived_courses = []
if split_archived:
    archived_courses = six.moves.filter(operator.methodcaller("has_ended"), active_courses)
    active_courses = six.moves.filterfalse(operator.methodcaller("has_ended"), active_courses)
```